FIPS/fip-draft_foundational_fvm.md

```diff
@@ -0,0 +1,1010 @@
+  - [Memory policy](#memory-policy)
+  - [Module caching](#module-caching)
+  - [Gas charging](#gas-charging)
+  - [Specific actor coupling](#specific-actor-coupling)
+- [Foundational FVM network upgrade](#foundational-fvm-network-upgrade)
+- [Design Rationale](#design-rationale)
+- [Backwards Compatibility](#backwards-compatibility)
+- [Test Cases](#test-cases)
+- [Security Considerations](#security-considerations)
+- [Incentive Considerations](#incentive-considerations)
+- [Product Considerations](#product-considerations)
+- [Implementation](#implementation)
+- [Copyright](#copyright)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Simple Summary
+
+User-programmability is the ability to deploy and run arbitrary user-provided
+code on some infrastructure, in this case a blockchain network.
+
+Today, the Filecoin blockchain lacks user-programmability. This FIP is part of a
+series that aims to introduce this capability.
+
+A change of this magnitude impacts many components. Concretely:
+
+1. The storage layer, to store and load user-provided code.
+2. The execution layer, to compile/interpret user-provided code, run state
+   transitions involving such code, and verify the results.
+3. The gas model, to secure the network and modulate chain capacity
+   demand/supply dynamics in the face of user-programmable code.
+4. The preexisting baseline protocol, to facilitate programmability against
+   system functionality.
+
+This FIP focuses on (1) and (2). The remaining points are out of scope of this
+FIP, and will be addressed by subsequent proposals.
+
+Moreover, this FIP proposes a network upgrade to introduce a foundational
+version of the FVM.
+
+## Abstract
+
+We specify the Filecoin Virtual Machine (FVM), a IPLD-ready WASM-based execution
+layer capable of running arbitrary user-provided code. The FVM replaces the
+existing execution layer.
+
+The logic of native actors (the equivalent of smart contracts in Filecoin) is
+deployed as WASM bytecode. Foreign actors targeting non-Filecoin runtimes such
+as the Ethereum Virtual Machine are also supported through runtime emulation.
+
+We provide a reference architecture for the FVM, and explain its various
+components: (a) Machine, (b) Call Manager, (c) Invocation Container, (d) Kernel,
+(e) Syscalls, and (f) Externs. We define the contract that actor code must
+adhere to, the catalogue of available syscalls, the mechanics of state
+manipulation, the error conditions, and more technical details.
+
+In this FIP, we propose a network upgrade involving two major changes: (1) the
+introduction of a non-programmable version of the FVM to pave the way, and (2)
+gas model changes. The latter is not specified herein, and is object of an
+upcoming companion FIP.
+
+## Change Motivation
+
+The desire for user-programmability in Filecoin is well-recognized and
+uncontested. It is the key to unlock enormous latent potential, as it enables
```

Comment:
I wouldn't say "uncontested", that's inviting someone to challenge you.

Reply:
Ok, I'll lower the tone :-)